File: js/src/pdf.ts
import { type Engine, PartHTML } from "./guten.js";

/**
 * PdfConverter converts rendered HTML to PDF bytes. HTML -> PDF needs a
 * rendering engine (puppeteer / headless Chromium), a system-level dependency,
 * so guten keeps it an injectable seam rather than a core dependency — the
 * consumer provides the converter (see the README for adapter notes).
 */
export interface PdfConverter {
  toPdf(html: string): Promise<Uint8Array> | Uint8Array;
}

/** renderToPdf renders the template's html part and converts it via converter. */
export async function renderToPdf(
  engine: Engine,
  name: string,
  data: Record<string, unknown>,
  converter: PdfConverter,
): Promise<Uint8Array> {
  if (!converter) throw new Error("guten: no PDF converter provided");
  const html = engine.renderPart(name, PartHTML, data);
  return converter.toPdf(html);
}

File: go/pdf.go
package guten

import (
	"context"
	"fmt"
)

// PDFConverter converts rendered HTML to PDF bytes.
//
// HTML -> PDF requires a rendering engine (headless Chromium via chromedp, or
// wkhtmltopdf) — a system dependency — so guten keeps it an injectable seam
// rather than a core dependency, preserving the engine's pure/offline nature.
// The consumer provides the converter; see the README for adapter notes.
type PDFConverter interface {
	ToPDF(ctx context.Context, html []byte) ([]byte, error)
}

// RenderToPDF renders the template's html part and converts it to PDF via the
// provided converter.
func (e *Engine) RenderToPDF(ctx context.Context, name string, data map[string]any, converter PDFConverter) ([]byte, error) {
	if converter == nil {
		return nil, fmt.Errorf("guten: no PDF converter provided")
	}
	htmlBody, err := e.RenderPart(name, PartHTML, data)
	if err != nil {
		return nil, err
	}
	return converter.ToPDF(ctx, []byte(htmlBody))
}
